File: php/lox/Lox.php
<?php

declare(strict_types=1);

namespace Lox;

class Lox
{
    private static $hadError = false;

    /**
     * @param string[] $args
     * @return void
     */
    public static function main(
        array $args
    ): void
    {
        if (count($args) > 2) {
            echo ('Usage: jlox [script]');
            exit(64);
        } else if (count($args) == 2) {
            self::runFile($args[0]);
        } else {
            self::runPrompt();
        }
    }

    private static function runFile(string $path): void
    {
        self::run(file_get_contents($path));
        if (self::$hadError) exit(65);
    }

    private static function runPrompt()
    {
        $input = STDIN;

        for (;;) {
            echo '> ';
            $line = fgets($input);
            if ($line === false) break;
            self::run($line);
            self::$hadError = false;
        }
    }

    private static function run(string $source): void
    {
        $scanner = new Scanner($source);
        $tokens = $scanner->scanTokens();

        foreach ($tokens as $token) {
            echo $token . "\n";
        }
    }

    static function error(int $line, string $message): void
    {
        self::report($line, "", $message);
    }

    private static function report(
        int $line,
        string $where,
        string $message
    ): void
    {
        fputs(STDERR, "[line $line] Error$where: $message");
        self::$hadError = true;
    }
}

File: go/lox/lox.go
package lox

import (
	"bufio"
	"fmt"
	"os"

	"github.com/AsazuTaiga/crafting_interpriters/go/interpreter"
	"github.com/AsazuTaiga/crafting_interpriters/go/logger"
	"github.com/AsazuTaiga/crafting_interpriters/go/parser"
	"github.com/AsazuTaiga/crafting_interpriters/go/scanner"
)

type Lox struct {
	logger *logger.Logger
	interpreter *interpreter.Interpreter
}

func NewLox(
	logger *logger.Logger,
) *Lox {
	i := interpreter.NewInterpreter()
	return &Lox{
		logger: logger,
		interpreter: i,
	}
}

func (l *Lox) Run() {
	args := os.Args[2:]

	if len(args) > 1 {
		fmt.Println("Usage: golox [script]")
		os.Exit(64)
	} else if len(args) == 1 {
		l.runFile(args[0])
	} else {
		l.runPrompt()
	}
}

func (l *Lox) runFile(path string) error {
	bytes, err := os.ReadFile(path)
	if err != nil {
		return err
	}

	l.run(string(bytes))
	if(l.logger.HadError()) {
		os.Exit(65)
	}
	return nil
}

func (l *Lox) runPrompt() {
	reader := bufio.NewReader(os.Stdin)

	for {
		fmt.Print("> ")
		line, err := reader.ReadString('\n')
		if err != nil {
			break
		}

		l.run(line)
		l.logger.ResetError()
	}
}

func (l *Lox) run(source string) {
	s := scanner.NewScanner(source)
	tokens := s.ScanTokens(l.logger)
	parser := parser.NewParser(tokens)
	expr := parser.Parse()

	if l.logger.HadError() {
		return
	}


	l.interpreter.Interpret(expr)
}
